torch/distributed/rpc/backend_registry.py

```diff
@@ -277,38 +295,88 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
+            name,
+            device_count,
+            rpc_backend_options.device_maps,
+            rpc_backend_options.devices,
+            group,
+        )
 
-    # TODO: add try-except and destroy _agent in all processes if any fails.
-    agent = TensorPipeAgent(
-        store,
-        name,
-        rank,
-        world_size,
-        rpc_backend_options,
-        reverse_device_maps,
-        devices,
-    )
+        # TODO: add try-except and destroy _agent in all processes if any fails.
+        agent = TensorPipeAgent(
+            store,
+            name,
+            rank,
+            world_size,
+            rpc_backend_options,
+            reverse_device_maps,
+            devices,
+        )
 
-    api._init_rpc_states(agent)
+        api._init_rpc_states(agent)
 
-    # Run one dummy round of RPC to initialize channels/transports. Without
-    # this, it's easy to hit timeout in rpc.shutdown() if there is no other RPC
-    # on that process before rpc.shutdown(), as the agent initialization can
-    # take longer than 5s.
-    api._all_gather(None, timeout=rpc_backend_options.rpc_timeout)
-    # Need a barrier here to make sure no peers leave before the rank0 finishes
-    # _all_gather
-    group.barrier().wait()
+        # Run one dummy round of RPC to initialize channels/transports. Without
+        # this, it's easy to hit timeout in rpc.shutdown() if there is no other RPC
+        # on that process before rpc.shutdown(), as the agent initialization can
+        # take longer than 5s.
+        api._all_gather(None, timeout=rpc_backend_options.rpc_timeout)
+        # Need a barrier here to make sure no peers leave before the rank0 finishes
+        # _all_gather
+        group.barrier().wait()
 
-    return agent
+        return agent
+    # initialization for dynamic rpc (ranks can join and leave)
+    else:
+        # Validate devices and device_maps locally for current rank
+        _tensorpipe_check_local_device_maps(name, device_count, rpc_backend_options)
+
+        token_key = "RpcGroupManagementToken"
+        token_location = f"TokenOnWorker{rank}"
+        while True:
+            # Retrieve token from store to signal start of rank join/leave critical section
+            returned = store.compare_set(token_key, "", token_location).decode()
+            if returned == token_location:
+                # Construct TPAgent with empty reverse_device_map and devices
+                # these two properties will be updated after construction
+                agent = TensorPipeAgent(
+                    store,
+                    name,
+                    rank,
+                    world_size,
+                    rpc_backend_options,
+                    {},
+                    [],
+                )
 
+                try:
+                    # TODO: Notify all workers in group this rank has joined and set devices and reverse_device_map
```

Comment:
Looks like without this, we cannot reasonable coverage on tests. Let's wait till this is addressed in the next PR before landing.

Reply:
I added tests in `rpc_test` to cover initialization. There will be tests in remaining PRs which enables communication and can be tested.